### ISSUE TITLE
TempFanTower script not taking temperature changes into account that are already in the Gcode

### ISSUE THREAD

**Issue by DBMcoder** [5axes/Cura-Postprocessing-Scripts#14]
I'm using:
Windows 11Pro 23H2
Cura 5.7.0
Ultimaker S5 gen1 printer with material station
_Edit: 
I'm using  AA 0,8mm printcores in both extruders and a 0,2mm layer height.
I've set the script settings to: 
- offset layer 8
- start with 265 degrees C
- increments from 5 degrees C
- change at layerheight 50_

I used this [model ](https://www.thingiverse.com/thing:2729076) to perform a temperature test for PETG filament. I ran the post processing script: TempFanTower and set the settings...

I ran the test and noticed several unwanted temperature changes:
![temp tower temp profile](https://github.com/5axes/Cura-Postprocessing-Scripts/assets/81773252/0da12d6c-b1cd-4512-92f8-978b9d95a3d2)

I unpacked the *.UFP file and looked at the Gcode and saw the added temperature changes as 'M140 S' but also other temperature changes also M104 S, probably generated by Cura (could't find out why).

I looked at the TempFanTower.py script on github and noticed there isn't a routine that does a check if there are more M104 S codes in between the added temperature settings for every layer change. It should do that. 

Furthermore it's better to use the M109 R command; this waits for the nozzle to reach the setted temperature before moving on and it does this for both cooling off or heating up to that temperature (see [here](https://marlinfw.org/docs/gcode/M109.html))

**Comment by 5axes** [comment]
Hello,
Thanks for this report. In Cura you should not have different Nozzle value. So yes in the script there is certainly no check for other M104 code. 
For the M109 on my point of view it's not necessar , the temperature modification is very fast when you decrease the value so no needs to wait the right value.
But you are free to modify this script for your own needs.

**Comment by DBMcoder** [comment]
@5axes: Can you explain what you mean with: "should not have different Nozzle value"?  
And to comment further on your reply: Yes I understand that I am free to modify the script for my own needs :-) **But I wanted to raise the issue that the temp tower script in my use case is not working correctly**. I was lucky to check the actual temperatures settings during printing so I noticed, but otherwise I wouldn't have noticed and looked at an temperature tower that is not actually a temperature tower. The temperature it consistently goes back to is at every bridge part, where it is actually interesting to see what happens to it with different temperature settings. So all my bridges are printed at exactly the same temperature (232 degrees C, the original temperature setting of the material)

So I'm wondering if more people see this behaviour and if so, the script should be changed in my opinion to comment out or delete all M104 temperature settings that deviate from the script setting.

**Comment by 5axes** [comment]
Yes, but theoretically there are no temperature variations in the files generated by Cura (except for the first layer). 

You should find out why in your case there are M104 S codes in the base file.

**Comment by DBMcoder** [comment]
Yes indeed I'm currently investigating this. 

I did some settings variations and compared the generated gcode from the .ufp files. 
I used the 'balanced extra fast' profile in Cura, still with 0.8mm AA printcores and PETG

With all the settings unchanged the gcode doesn't generate any extra nozzle temp settings, so the script works. Temptower got printed correctly (woehooe) But when I change all the print speeds to 75mm/s, nozzle temperature changes within a layer are introduced in the gcode. Now there are 243 M104 temperature changes......

I have no clue why, my guess is that there is a calculation somewhere that prevents that the power of the printcore is maxed out (which is pretty low; only 20 Watts I believe) and that the strategy is to lower the temperature in favour of keeping the printspeed? I'll post this on the Ultimaker forum to find out more.....

So if there is such a calculation done in Cura, it does can happen to anyone that M104 codes are introduced. And since the tempfantower is a test to find the right print settings (finding the right temperature for a material with a certain nozzle size and print speed) it is definitely possible that people will play with speed settings....

**Comment by 5axes** [comment]
Yes same "issue" with the Print Core AA 0.8

A Temperature variation between 223° & 225°C
![image](https://github.com/5axes/Cura-Postprocessing-Scripts/assets/11015345/1f933803-2eed-4537-be0a-f55ebd89e3ae)

**Comment by 5axes** [comment]
Ok I think it's due to the hidden parameter :  **Flow Temperature Graph**

![image](https://github.com/5axes/Cura-Postprocessing-Scripts/assets/11015345/3f71bdae-3bca-4642-a23e-ff5684f068ae)


Not actif on the majority of the printer definition in Cura ...

**Comment by DBMcoder** [comment]
So that looks like this is implemented for the power limit of the printcore, since flow and temperature lead to power. I did a max flow test with ABS with an 0.8mm AA core, 0,3mm layer height and 262 degrees C printtemp at 100mm/2 where I got an ER31 fault code, maxing the power of the heat core, so it couldn't keep the temperature up. Green line is heater duty cycle maxing at 255
![maxflow test ABS cards3D hogere printtemp - failed](https://github.com/5axes/Cura-Postprocessing-Scripts/assets/81773252/e9d1caf6-2e5a-475d-9444-d4c1a39f549b)
